crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+        return 0;
+    }
+
+    if ((tmpl->extensions) == NULL) {
+        if ((tmpl->extensions = sk_X509_EXTENSION_new_null()) == NULL)
+            goto oom;
+        new = 1;
+    }
+
+    if (!sk_X509_EXTENSION_push(tmpl->extensions, (X509_EXTENSION *)ext))
+        goto oom;
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_EXTENSION, ERR_R_MALLOC_FAILURE);
+
+    if (new != 0) {
+        sk_X509_EXTENSION_free(tmpl->extensions);
+        tmpl->extensions = NULL;
+    }
+    return 0;
+}
+
+/* TODO: support cases 1+2 (besides case 3) defined in RFC 4211, section 4.1. */
+static int CRMF_poposigningkey_init(OSSL_CRMF_POPOSIGNINGKEY *ps,
+                                    OSSL_CRMF_CERTREQUEST *cr,
+                                    const EVP_PKEY *pkey, int dgst)
+{
+    int len;
+    size_t crlen, max_sig_size;
+    unsigned int siglen;
+    unsigned char *crder = NULL, *sig = NULL;
+    int alg_nid=0, md_nid=0;
+    const EVP_MD *alg = NULL;
+
+    EVP_MD_CTX *ctx = NULL;
+
+    if (ps == NULL || cr == NULL || pkey == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    ps->signature->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+
+    len = i2d_OSSL_CRMF_CERTREQUEST(cr, &crder);
+    if (len < 0 || crder == NULL)
+        goto err;
+    crlen = (size_t)len;
+
+    max_sig_size = EVP_PKEY_size((EVP_PKEY *)pkey);
+    sig = OPENSSL_malloc(max_sig_size);
+    if (sig == NULL)
+        goto err;
+
+    if (!OBJ_find_sigid_by_algs(&alg_nid, dgst, EVP_PKEY_id(pkey))) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!(OBJ_find_sigid_algs(alg_nid, &md_nid, NULL)
+          && (alg = EVP_get_digestbynid(md_nid)) != NULL)) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT,
+                CRMF_R_UNSUPPORTED_ALG_FOR_POPSIGNINGKEY);
+        goto err;
+    }
+    if (!X509_ALGOR_set0(ps->algorithmIdentifier, OBJ_nid2obj(alg_nid),
+                         V_ASN1_NULL, NULL)
+        || (ctx = EVP_MD_CTX_new()) == NULL
+        || !(EVP_SignInit_ex(ctx, alg, NULL)))
+        goto err;
+    if (!(EVP_SignUpdate(ctx, crder, crlen)))
+        goto err;
+    if (!(EVP_SignFinal(ctx, sig, &siglen, (EVP_PKEY *)pkey)))
+        goto err;
+
+    if (!(ASN1_BIT_STRING_set(ps->signature, sig, siglen)))
+        goto err;
+
+    /* cleanup */
+    OPENSSL_free(crder);
+    EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 1;
+ err:
+    CRMFerr(CRMF_F_CRMF_POPOSIGNINGKEY_INIT, CRMF_R_ERROR);
+    OPENSSL_free(crder);
+    if (ctx)
+        EVP_MD_CTX_free(ctx);
+    OPENSSL_free(sig);
+    return 0;
+}
+
+
+int OSSL_CRMF_MSG_create_popo(OSSL_CRMF_MSG *crm, const EVP_PKEY *pkey,
+                              int dgst, int ppmtd)
```

Comment:
What is `dgst` here? Should this really be an `EVP_MD *`?

Reply:
It's the NID of the desired digest. We found it more convenient to pass around than `EVP_MD *`.